mysql/provider.go

```diff
@@ -359,3 +360,20 @@ func mysqlErrorNumber(err error) uint16 {
 	}
 	return me.Number
 }
+
+func cloudsqlErrorNumber(err error) int {
+	if err == nil {
+		return 0
+	}
+
+	if e, ok := err.(*googleapi.Error); ok {
```

Comment:
What's the point of this code? Isn't it so implementation-specific, that it will break once some internals of cloudsql change?

Especially: why do we check for Code to be in 400s when "if" is true and not in the `else`?

Your whole function could be likely written like this without external libraries (it could contain the check for 400s, if that is important):
```
import "errors"
...
var gapiError *googleapi.Error
if errors.As(err, &gapiError) {
  return gapiError.Code
}
return 0
```

Reply:
You're right these code is much better